api/src/services/sso-login.js

```
@@ -0,0 +1,210 @@
+ * 
+ * @param {string} currentUrl Current url that contains authorization code grant.
+ * @returns {object} Token id and user details.
+ */
+const getIdToken = async (currentUrl) => {
+
+  try {
+    const serverConfig = await oidcServerSConfig();
+    const tokens = await networkCallRetry(
+      () => client.authorizationCodeGrant(serverConfig, currentUrl, { idTokenExpected: true })
+    );
+    const { id_token } = tokens;
+    const { name, preferred_username, email } = tokens.claims();
+
+    return {
+      id_token,
+      user: {
+        name,
+        username: preferred_username,
+        email
+      }
+    };
+  } catch (err) {
+    logger.error(err);
+    throw SERVER_ERROR;
+  }
+};
+
+/**
+ * Makes a CouchDB session cookie value string.
+ *  
+ * @param {string} username Username.
+ * @param {string} salt CouchDB user salt value.
+ * @param {string} secret  CouchDB secret value.
+ * @param {number} authTimeout Cooke timeout.
+ * @returns {string} Cookie string.
+ */
+const makeCookie = (username, salt, secret, authTimeout) => {
+  // an adaptation of https://medium.com/@eiri/couchdb-cookie-authentication-6dd0af6817da
+  const expiry = Math.floor(Date.now() / 1000) + authTimeout;
+  const msg = `${username}:${expiry.toString(16)
+    .toUpperCase()}`;
+  const key = `${secret}${salt}`;
+  const hmac = createHmac('sha1', key)
+    .update(msg)
+    .digest();
+  const control = new Uint8Array(hmac);
+  let cookie = Buffer.concat([
+    Buffer.from(msg),
+    Buffer.from(':'),
+    Buffer.from(control)
+  ]).toString('base64');
+
+  while (cookie.endsWith('=')) {
+    cookie = cookie.slice(0, -1);
+  }
+
+  return cookie
+    .replaceAll('/', '_')
+    .replaceAll('+', '-');
+};
+
+/**
+ * Gets a session cookie. 
+ * 
+ * @param {string} username Username.
+ * @returns {string} Session cookie.
+ */
+const getCookie = async (username) => {
+  let secret;
+  let authTimeout;
+  let salt;
+
+  try {
+    const userDoc = await users.getUserDoc(username);
```

If we are resolving users by `email`, we will not be able to use this function to get the user doc.  Instead, I am thinking we should update the [`users_by_field` view query](https://github.com/medic/cht-core/blob/61a04a62d19323f87b2020c62ee07ed5dbd56695/ddocs/users-db/users/views/users_by_field/map.js) to emit the users by their `email` value.  Then we can query here with code like:

```
    const [userDoc, ...additionalUsers] = await db.users.query(
      'users/users_by_field',
      {
        key: ['email', email],
        include_docs: true
      }
    ).then(({ rows }) => {
      return rows
        .map(({ doc }) => doc)
        .filter(doc => doc?.oidc);
    });
```

Then, we should throw an exception if `userDoc` is undefined OR if `additionalUsers` _is NOT empty_ (that means more than one user has this email value).

---

@benkags I found an issue with this proposed implementation. :disappointed: It turns out that we do not actually set the user's `email` value onto their `_users` document.  Instead that gets stored in their `user-settings` document in the `medic` database.  

My new proposal is that we drop the email stuff from this PR and just wrap this up by keeping the `preferred_username` mapping for now as you have it here.   Then, I can follow up with another PR for https://github.com/medic/cht-core/issues/9890 that can add a new Couch View index for looking up the `user-settings` by email.

---

Thanks for looking into the email support in `_users` objects. Agree on wrapping this up with current `preferred_username` approach.